### ISSUE TITLE
did not receive `didSelect` calls on my `MGLMapViewDelegate` delegate

### ISSUE THREAD

**Issue by bugaco** [maplibre/maplibre-gl-native#187]
Hi, 
I just create some Annotation on the map, and set the mapView's delegate to the Controller, but can't receive `didSelect` method calls.

This issue looks maybe simple, but I can't figure it out after hours passing by🥺

This is demo code, 
and did not receive "didSelect" in Console when tapping  annotation:
```Swift
class AnnotationViewController: UIViewController, MGLMapViewDelegate {
    
    override func viewDidLoad() {
        super.viewDidLoad()

        let mapView = MGLMapView(frame: view.bounds)
        mapView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        mapView.styleURL = URL(string: "https://api.maptiler.com/maps/streets/style.json?key=JXULcy1Vdmf0O6K1iqbM")
        mapView.tintColor = .lightGray
        mapView.centerCoordinate = CLLocationCoordinate2D(latitude: 0, longitude: 66)
        mapView.zoomLevel = 2
        mapView.delegate = self
        view.addSubview(mapView)

        // Specify coordinates for our annotations.
        let coordinates = [
            CLLocationCoordinate2D(latitude: 0, longitude: 33),
            CLLocationCoordinate2D(latitude: 0, longitude: 66),
            CLLocationCoordinate2D(latitude: 0, longitude: 99)
        ]

        // Fill an array with point annotations and add it to the map.
        var pointAnnotations = [MGLPointAnnotation]()
        for coordinate in coordinates {
            let point = MGLPointAnnotation()
            point.coordinate = coordinate
            point.title = "\(coordinate.latitude), \(coordinate.longitude)"
            pointAnnotations.append(point)
        }

        mapView.addAnnotations(pointAnnotations)
    }

    // MARK: - MGLMapViewDelegate methods

    // This delegate method is where you tell the map to load a view for a specific annotation. To load a static MGLAnnotationImage, you would use `-mapView:imageForAnnotation:`.
    func mapView(_ mapView: MGLMapView, viewFor annotation: MGLAnnotation) -> MGLAnnotationView? {
        // This example is only concerned with point annotations.
        guard annotation is MGLPointAnnotation else {
            return nil
        }

        // Use the point annotation’s longitude value (as a string) as the reuse identifier for its view.
        let reuseIdentifier = "\(annotation.coordinate.longitude)"

        // For better performance, always try to reuse existing annotations.
        var annotationView = mapView.dequeueReusableAnnotationView(withIdentifier: reuseIdentifier)

        // If there’s no reusable annotation view available, initialize a new one.
        if annotationView == nil {
            annotationView = CustomAnnotationView(reuseIdentifier: reuseIdentifier)
            annotationView!.bounds = CGRect(x: 0, y: 0, width: 40, height: 40)

            // Set the annotation view’s background color to a value determined by its longitude.
            let hue = CGFloat(annotation.coordinate.longitude) / 100
            annotationView!.backgroundColor = UIColor(hue: hue, saturation: 0.5, brightness: 1, alpha: 1)
        }

        return annotationView
    }

    func mapView(_ mapView: MGLMapView, annotationCanShowCallout annotation: MGLAnnotation) -> Bool {
        return true
    }
    
    func mapView(_ mapView: MGLMapView, shapeAnnotationIsEnabled annotation: MGLShape) -> Bool {
        true
    }
    
    func mapView(_ mapView: MGLMapView, didSelect annotationView: MGLAnnotationView) {
        print("didSelect annotationView")
    }
    
    func mapView(_ mapView: MGLMapView, didSelect annotation: MGLAnnotation) {
        print("didSelect annotation")
    }
}

//
// MGLAnnotationView subclass
class CustomAnnotationView: MGLAnnotationView {
    override func layoutSubviews() {
        super.layoutSubviews()

        // Use CALayer’s corner radius to turn this view into a circle.
        layer.cornerRadius = bounds.width / 2
        layer.borderWidth = 2
        layer.borderColor = UIColor.white.cgColor
    }

    override func setSelected(_ selected: Bool, animated: Bool) {
        super.setSelected(selected, animated: animated)

        // Animate the border width in/out, creating an iris effect.
        let animation = CABasicAnimation(keyPath: "borderWidth")
        animation.duration = 0.1
        layer.borderWidth = selected ? bounds.width / 4 : 2
        layer.add(animation, forKey: "borderWidth")
    }
}

```

**Comment by dieterdreist** [comment]
I have the same issue (maybe). https://github.com/maplibre/maplibre-gl-native/issues/176
at least ti me it seems that the select is not issued because visibleAnnotations is nil and the code checks against visibleAnnotations when determining if a map tap is on an annotation.

**Comment by bugaco** [comment]
> I have the same issue (maybe). #176
> at least to me it seems that the _select_ is not issued because visibleAnnotations is nil and the code checks against visibleAnnotations when determining if a map tap is on an annotation.

I also have the issue you mentioned in #176. 
It's odd, I just see the annotations on the map, but `visibleAnnotations` is nil.🤯

Inspired by your replay, I checked the `visibleAnnotations` property, and found that:  
When annotation did not response to `didSelect`, the mapView's `visibleAnnotations` property is nil.
When `visibleAnnotations` is not nil, MapViewDelegate can receive some `didSelect` calls.

So, I consider this two issue is caused by the the same bug(may be).

**Comment by mehlkelm** [comment]
Same issue here since 5.12.1 (worked before).
The following ViewController Setup reproduces the problem. With 5.12.0, taps on the annotation are correctly logged to the console, with 5.12.1 they are missing.

`import UIKit
import Mapbox

class ViewController: UIViewController, MGLMapViewDelegate {

    override func viewDidLoad() {
        super.viewDidLoad()
        
        let mapView = MGLMapView(frame: view.bounds, styleURL: Bundle.main.url(forResource: "maptiler-demo-style", withExtension: "json")!)
        mapView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        mapView.delegate = self
        view.addSubview(mapView)
    }
    
    func mapViewDidFinishLoadingMap(_ mapView: MGLMapView) {
        let test = MyAnnotation(coordinate: CLLocationCoordinate2D(latitude: 47.356, longitude: 8.530))
        mapView.addAnnotations([test])
        mapView.setCenter(test.coordinate, zoomLevel: 13, animated: true)
    }
    
    func mapView(_ mapView: MGLMapView, viewFor annotation: MGLAnnotation) -> MGLAnnotationView? {
        if annotation is MyAnnotation {
            let view = MyAnnotationView(annotation: annotation, reuseIdentifier: "AnnotationView")
            return view
        }
        return nil
    }
    
    func mapView(_ mapView: MGLMapView, didSelect annotation: MGLAnnotation) {
        
        if annotation is MGLUserLocation {
            print("Did select USER annotation")
        } else {
            print("Did select annotation")
        }
    }
    
    func mapView(_ mapView: MGLMapView, didSelect annotationView: MGLAnnotationView) {
        
        print("Did select annotation view")
    }
}

class MyAnnotation: NSObject, MGLAnnotation {
    var ID = UUID()
    var coordinate: CLLocationCoordinate2D

    init(coordinate: CLLocationCoordinate2D) {
        self.coordinate = coordinate
        super.init()
    }
}

class MyAnnotationView: MGLAnnotationView {
    override init(annotation: MGLAnnotation?, reuseIdentifier: String?) {
        super.init(annotation: annotation, reuseIdentifier: reuseIdentifier)
        let size = CGSize(width: 40, height: 40)
        let origin = CGPoint(x: 0, y: 0)
        frame = CGRect(origin: origin, size: size)

        let image = UIImageView(image: UIImage(systemName: "doc.fill"))
        image.tintColor = .blue
        image.frame = CGRect(origin: .zero, size: size)
        addSubview(image)
    }
        
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
}`

**Comment by fonok3** [comment]
Does this [release](https://github.com/maplibre/maplibre-gl-native/releases/tag/ios-v5.12.2) fixes the issues? It fixes a similar issue with queried features (#147). It also worked at 5.12.0 and was broken in 5.12.1.

**Comment by mehlkelm** [comment]
> Does this [release](https://github.com/maplibre/maplibre-gl-native/releases/tag/ios-v5.12.2) fixes the issues? It fixes a similar issue with queried features (#147). It also worked at 5.12.0 and was broken in 5.12.1.

It seems to be fixed with 5.12.2, although I can only verify on iOS 14 Simulator due to #268
